R/getMappableRegions.R

```diff
@@ -0,0 +1,189 @@
+#'     be printed on the console.
+#'
+#' @details Sequences of all overlapping windows are extracted from the genome
+#'     and aligned back to that genome using \code{\link[Rbowtie]{bowtie}} with
+#'     parameters \code{--concise -f -v 0 -a -B 1 -m maxHits}. If no more than
+#'     \code{maxHits} hits are found, the window is defined mappable.
+#'
+#' @return A \code{\link[GenomicRanges]{GRanges}} object with mappable regions.
+#'     All plus-strand sequences in \code{genome} of length \code{kmerLength}
+#'     with their start (leftmost) position overlapping the \code{GRanges} object
+#'     do not generate more than \code{maxHits} hits when aligned to the genome.
+#'
+#' @examples
+#' library(Rbowtie)
+#' 
+#' genomefile <- system.file("extdata", "getMappableRegions", "hg19sub.fa", package = "swissknife")
+#' indexdir <- tempfile()
+#' indexpre <- "index"
+#' indexname <- file.path(indexdir, indexpre)
+#' idx <- bowtie_build(genomefile, indexdir)
+#' 
+#' mapgr <- getMappableRegions(genomefile, indexname, 50, quiet = FALSE)
+#' mapgr
+#'
+#' @seealso \code{\link[Rbowtie]{bowtie}} in package \pkg{Rbowtie} used by
+#'     \code{getMappableRegions} to align reads to the genome;
+#'     \code{\link[Rbowtie]{bowtie_build}} in package \pkg{Rbowtie} for
+#'     indexing a genome.
+#'
+#' @importFrom GenomicRanges GRanges gaps sort
+#' @importFrom Biostrings readDNAStringSet
+#' @importFrom BSgenome seqinfo getSeq
+#' @importFrom GenomeInfoDb seqnames seqlengths
+#' @importFrom XVector width
+#' 
+#' @export
+getMappableRegions <- function(genome,
+                               genomeIndex,
+                               kmerLength = 50,
+                               maxHits = 1,
+                               Ncpu = 2,
+                               quiet = TRUE) {
+    # check arguments
+    # ... genome
+    if(is(genome, "BSgenome")) {
+        chrinfo <- BSgenome::seqinfo(genome)
+        chrs <- BSgenome::getSeq(genome, names = seqnames(chrinfo),
+                                 start = 1, end = seqlengths(chrinfo))
+    } else if (is.character(genome) && length(genome) == 1L) {
+        if (suppressWarnings(require(genome, character.only = TRUE, quietly = TRUE))) {
+            genome <- get(genome)
+            chrinfo <- BSgenome::seqinfo(genome)
+            chrs <- BSgenome::getSeq(genome, names = seqnames(chrinfo),
+                                     start = 1, end = seqlengths(chrinfo))
+        } else if (file.exists(genome)) {
+            chrs <- Biostrings::readDNAStringSet(genome)
+        } else {
+            stop("'genome' is neither a valid file nor a BSgenome object.")
+        }
+    }
+    if (!quiet) {
+        message("Will calculate mappability for ", length(chrs), " chromosomes ",
+                "(", sum(width(chrs)), " base pairs)")
+    }
+    # ... genomeIndex
+    if (is.character(genomeIndex) && length(genomeIndex) == 1L) {
```

Comment:
I wonder if there is any way we can check for likely (in)compatibilities between the genome or the index (or perhaps we just have to leave this to the user)

Reply:
Good point. The easy way to check would be using `bowtie-inspect` and compare the chromosome sizes. **Rbowtie** does not included `bowtie-inspect`, so as an alternative one could start an alignment and check the `@SQ` lines in the sam header?

Reply:
Ok, I have added a check along those lines (not perfect, but at least chromosome names and lengths need to be consistent)

Reply:
Nice! I think this is enough to be safe from all realistic mistakes.